### PR TITLE
Added support for VTK

### PR DESCRIPTION
**Pull request by pcafrica**
After a discussion with @luca-heltai and @fdrmrc, we would like to add support for the [VTK](https://vtk.org/) library in deal.II.

VTK can be used for importing scalar or vector fields, evaluate/interpolate/project them onto a triangulation and many other geometry handling tasks. We plan to support versions higher or equal to 9.0.0. These features should complement the use of [CGAL](#13644).

VTK has already been extensively tested and used within [lifex](https://gitlab.com/lifex/lifex) and we aim to port existing functionalities from lifex to deal.II. I expect many more PR to follow once this is accepted.

TODO:
- [x] check workflow file: is it ok to update to Ubuntu 22.04?
- [x] add sample class
- [x] add tests

### CODE REVIEW

**Comment by bangerth** [comment]
I'll let @tamiko do the review.

Out of curiosity, what features in VTK do you plan on using?

**Comment by pcafrica** [comment]
> I'll let @tamiko do the review.
> 
> Out of curiosity, what features in VTK do you plan on using?

Most importantly, reading a (surface or volume) field from a `VTK` file and interpolating/projecting it at triangulation nodes using closest-point distance, linear interpolation or signed distance. This can be used for, _e.g._, imposing boundary conditions, reading a mesh displacement from file, defining different material properties or defining an immersed surface.

This will involve VTK filters like cell and point locators. An example can be found [here](https://gitlab.com/lifex/lifex/-/blob/main/source/io/vtk_function.hpp), where, inheriting from `Function<dim>`, we are able to evaluate an input field at an arbitrary triangulation point.

**Comment by bangerth** [comment]
Ah, very cool -- something like `FEFieldFunction`, just with externally provided data! That is something that I'm sure we can also put to good use in ASPECT, for example. I'm looking forward to the patch!

**Comment by agrayver** [comment]
Is there also a possibility here to export/import refined triangulations in a backward-compatible way? Current approach based on the  boost-serialization library is not backward compatible and breaks with each major release in practice (say meshes created with 9.3 cannot be read with 9.4 and so on...).

**Comment by pcafrica** [comment]
> Is there also a possibility here to export/import refined triangulations in a backward-compatible way? Current approach based on the boost-serialization library is not backward compatible and, in practice, breaks with each major release (say meshes created/saved with 9.3 cannot be read with 9.4 and so on...).

Personally I've never used VTK for this purpose, but it's something worth investigating for sure.

**Comment by pcafrica** [comment]
A note about the Linux workflow: installing `libvtk9-dev` from `apt` results in a dependency conflict that I wasn't able to solve: https://github.com/dealii/dealii/actions/runs/5073291921/jobs/9116267080?pr=15266#step:3:86

The current workaround is to build VTK manually, but this takes ~45 minutes and I guess it is too much (?).

**Comment by luca-heltai** [comment]
Can you try to add at the end

```
apt-get install ...

libvtk9-dev \
libtbb-dev-
```

(notice the `-` at the end of name of the missing package). I think we do install libtbb somewhere, since dealii depends on it. Maybe the naming is not right in libvtk9-dev.

**Comment by pcafrica** [comment]
> Can you try to add at the end
> 
> ```
> apt-get install ...
> 
> libvtk9-dev \
> libtbb-dev-
> ```
> 
> (notice the `-` at the end of name of the missing package). I think we do install libtbb somewhere, since dealii depends on it. Maybe the naming is not right in libvtk9-dev.

Thanks for the hint! Unfortunately, libvtk9-dev cannot be installed without libtbb-dev.

PS: `libtbb2-dev` is installed as a Trilinos dependency.

**Comment by luca-heltai** [comment]
My preference would be to add vtk as a dependency in the docker image dealii/dependencies:jammy (or focal) and use that one to build one of the workflow.

**Review by luca-heltai — APPROVED** [review]

**Comment by pcafrica** [comment]
Let's wait for #15278 to be discussed/merged before continuing with this PR.

**Review by luca-heltai — APPROVED** [review]

**Comment by luca-heltai** [comment]
Thanks Pasquale!

**Comment by luca-heltai** [comment]
Would you mind rebasing and squashing into a couple of significative commits?

**Comment by pcafrica** [comment]
> Would you mind rebasing and squashing into a couple of significative commits?

Done! From my side, it looks ready to merge.

**Comment by luca-heltai** [comment]
@tamiko, would you mind taking a look at the cmake stuff? I am fine with everything, but you are working on restructuring stuff....

**Review by luca-heltai — APPROVED** [review]

**Comment by luca-heltai** [comment]
Also, @pcafrica, would you mind adding `VTK` support also in `contrib/docker/Dockerfile`? You did this already in the `docker-files` repository, but this one is used by the master workflow to create docker images of master after each merge to master.

**Comment by pcafrica** [comment]
Done!

PS: the pipeline [fails](https://github.com/dealii/dealii/actions/runs/5106334018/jobs/9178472617#step:4:19) with error `mkdir: cannot create directory ‘build’: Permission denied`, do you have any clue about this @luca-heltai?

**Comment by luca-heltai** [comment]
Ah, yes. You have to run the pipeline as user root.

**Comment by luca-heltai** [comment]
I think we also need to set 

```
container:
      image: dealii/dependencies:jammy
      options: -u root -e OMPI_ALLOW_RUN_AS_ROOT=1 -e OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
```

**Comment by masterleinad** [comment]
Please fix the merge conflicts.

**Review by tamiko — APPROVED** [review]

File: cmake/modules/FindDEAL_II_VTK.cmake
Comment (tamiko) [inline cmake/modules/FindDEAL_II_VTK.cmake:55]: Let us merge this for now
We should support simply listing all targets now, though - but I suggest we try this out in a follow-up pull request.

**Pull request merged**
